thriftpy2/protocol/apache_json.py

```diff
@@ -0,0 +1,280 @@
+"""
+from __future__ import absolute_import
+import json
+import base64
+
+from six import string_types
+from thriftpy2.protocol import TProtocolBase
+from thriftpy2.thrift import TType
+
+
+CTYPES = {
+    TType.BOOL: 'tf',
+    TType.BYTE: 'i8',
+    TType.I16: 'i16',
+    TType.I32: 'i32',
+    TType.I64: 'i64',
+    TType.DOUBLE: 'dbl',
+    TType.STRING: 'str',
+    TType.BINARY: 'str',  # apache sends binary data as base64 encoded
+    TType.STRUCT: 'rec',
+    TType.LIST: 'lst',
+    TType.SET: 'set',
+    TType.MAP: 'map',
+}
+
+JTYPES = {v: k for k, v in CTYPES.items()}
+
+VERSION = 1
+
+
+def flatten(suitable_for_isinstance):
+    """
+    isinstance() can accept a bunch of really annoying different types:
+        * a single type
+        * a tuple of types
+        * an arbitrary nested tree of tuples
+    Return a flattened tuple of the given argument.
+    """
+
+    types = list()
+
+    if not isinstance(suitable_for_isinstance, tuple):
+        suitable_for_isinstance = (suitable_for_isinstance,)
+    for thing in suitable_for_isinstance:
+        if isinstance(thing, tuple):
+            types.extend(flatten(thing))
+        else:
+            types.append(thing)
+    return tuple(types)
+
+
+class TApacheJSONProtocolFactory:
+    @staticmethod
+    def get_protocol(trans):
+        return TApacheJSONProtocol(trans)
+
+
+class TApacheJSONProtocol(TProtocolBase):
+    """
+    Protocol that implements the Apache JSON Protocol
+    """
+
+    def __init__(self, trans):
+        TProtocolBase.__init__(self, trans)
+        self._req = None
+
+    def _load_data(self):
+        data = b""
+        l_braces = 0
+        in_string = False
+        while True:
+            # read(sz) will wait until it has read exactly sz bytes,
+            # so we must read until we get a balanced json list in absence of knowing
+            # how long the json string will be
+
+            new_data = self.trans.read(1)
+            data += new_data
+            if new_data == b'"' and not data.endswith(b'\\"'):
+                in_string = not in_string
+            if not in_string:
+                if new_data == b"[":
+                    l_braces += 1
+                elif new_data == b"]":
+                    l_braces -= 1
+            if l_braces == 0:
+                break
+        if data:
+            self._req = json.loads(data.decode('utf8'))
+        else:
+            self._req = None
+
+    def read_message_begin(self):
+        if not self._req:
+            self._load_data()
+        return self._req[1:4]
+
+    def read_message_end(self):
+        pass
+
+    def skip(self, ttype):
+        pass
+
+    def write_message_end(self):
+        pass
+
+    def write_message_begin(self, name, ttype, seqid):
+        self.api = name
+        self.ttype = ttype
+        self.seqid = seqid
+
+    def write_struct(self, obj):
+        """
+        Write json to self.trans following apache style jsonification of `obj`
```

Comment:
Missing new line between comments and type annotations.

Reply:
Fixed ffb624c